controller/kubernetes_configmap_controller.go

```diff
@@ -0,0 +1,246 @@
+		return
+	}
+	for i := 0; i < workers; i++ {
+		go wait.Until(kc.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (kc *KubernetesConfigMapController) worker() {
+	for kc.processNextWorkItem() {
+	}
+}
+
+func (kc *KubernetesConfigMapController) processNextWorkItem() bool {
+	key, quit := kc.queue.Get()
+	if quit {
+		return false
+	}
+	defer kc.queue.Done(key)
+	err := kc.syncHandler(key.(string))
+	kc.handleErr(err, key)
+	return true
+}
+
+func (kc *KubernetesConfigMapController) handleErr(err error, key interface{}) {
+	if err == nil {
+		kc.queue.Forget(key)
+		return
+	}
+
+	if kc.queue.NumRequeues(key) < maxRetries {
+		kc.logger.WithError(err).Warnf("Error syncing configmap %v", key)
+		kc.queue.AddRateLimited(key)
+		return
+	}
+
+	kc.logger.WithError(err).Warnf("Dropping configmap %v out of the queue", key)
+	kc.queue.Forget(key)
+	utilruntime.HandleError(err)
+}
+
+func (kc *KubernetesConfigMapController) syncHandler(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "%v: fail to sync %v", kc.name, key)
+	}()
+
+	namespace, cfmName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	if err := kc.reconcileDefaultStorageClass(namespace, cfmName); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (kc *KubernetesConfigMapController) reconcileDefaultStorageClass(namespace, cfmName string) error {
+	// The default storageclass configmap must be in the same namespace as the current KubernetesConfigMapController
+	if namespace != kc.namespace {
+		return nil
+	}
+
+	storageCFM, err := kc.ds.GetConfigMap(kc.namespace, types.DefaultStorageClassConfigMapName)
+	if err != nil {
+		if datastore.ErrorIsNotFound(err) {
+			// The storageclass configmap was deleted therefore we delete the actual storageclass
+			err := kc.ds.DeleteStorageClass(types.DefaultStorageClassName)
+			if err != nil && !datastore.ErrorIsNotFound(err) {
+				return err
+			}
+			return nil
+		}
+		return err
+	}
+
+	storageclass, err := buildStorageClassManifestFromConfigMap(storageCFM)
+	if err != nil {
+		return err
+	}
+
+	existingSC, err := kc.ds.GetStorageClassRO(types.DefaultStorageClassName)
+	if err != nil && !datastore.ErrorIsNotFound(err) {
+		return err
+	}
+
+	if storageclassesHaveSameValues(existingSC, storageclass) {
+		return nil
+	}
+
+	err = kc.ds.DeleteStorageClass(types.DefaultStorageClassName)
+	if err != nil && !datastore.ErrorIsNotFound(err) {
+		return err
+	}
+
+	storageclass, err = kc.ds.CreateStorageClass(storageclass)
+	if err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			return nil
+		}
+		return err
+	}
+
+	kc.logger.Infof("Updated the default Longhorn storageclass: %v", storageclass)
+
+	return nil
+}
+
+func buildStorageClassManifestFromConfigMap(cfm *v1.ConfigMap) (*storagev1.StorageClass, error) {
+	data, ok := cfm.Data["storageclass.yaml"]
+	if !ok {
+		return nil, fmt.Errorf("cannot find storageclass.yaml inside the default storageclass configmap")
```

Comment:
How about using errors.New() instead?

Reply:
It's more common to always use `fmt.Errorf` instead of switch between the two, since `errors.New()` doesn't provide a formatted output.